src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3431,32 +3432,34 @@ internal Buffer(IEnumerable<TElement> source, bool queryInterfaces = true)
                     if (collection != null)
                     {
                         count = collection.Count;
-                        if (count > 0)
+                        if (count != 0)
                         {
                             items = new TElement[count];
                             collection.CopyTo(items, 0);
                         }
+                        goto storeArray;
```

Comment:
I liked the previous version that just touched the below code (optimizing away the null check on each iteration) much more than I like this version with the gotos.  It seems like we're adding complexity and code simply to avoid an additional null check / branch on each ToArray call, but there's already a whole bunch of casts and null checks and branches on these code paths.  Does this really amount to a measurable win?